### PR TITLE
function structures

### PR DESCRIPTION
**Pull request by ecavallo**
Given structures `S` and `T`, we get a structure `\ X -> (S X -> T X)`. (The already defined structures on `X -> S X` and `A -> S X` are more convenient in those special cases, so we should keep them around.)

### CODE REVIEW

**Comment by ecavallo** [comment]
This is ready for review, although #379 should be dealt with first.
- Structured equivalences for `\ X -> S X -> T X` in general.
- Separate definition for `S` equipped with a functorial action. The current structured eqs for `\ X -> A -> T X` and `\ X -> X -> T X` are special cases of this.
- Structured relations for `\ X -> S X -> T X` when `S` is "positive" (includes being functorial and some other conditions)

For structured equivalences, functoriality is not really the right hypothesis, but that can be dealt with in a future PR.

File: Cubical/Foundations/RelationalStructure.agda [hunk 1]
```diff
@@ -76,26 +88,122 @@ record SuitableStrRel (S : Type ℓ → Type ℓ') (ρ : StrRel S ℓ'') : Type
...
 
+-- Additional conditions for a "positive" notion of structured relation
+
+isDetransitiveStrRel : {S : Type ℓ → Type ℓ'} (ρ : StrRel S ℓ'') → Type _
+isDetransitiveStrRel {ℓ = ℓ} {S = S} ρ =
+  {X Y Z : Type ℓ}
+  (R : PropRel X Y ℓ) (R' : PropRel Y Z ℓ)
+  {sx : S X} {sz : S Z}
+  → ρ (compPropRel R R' .fst) sx sz
+  → ∥ Σ[ sy ∈ S Y ] ρ (R .fst) sx sy × ρ (R' .fst) sy sz ∥
```
Comment (mortberg) [inline Cubical/Foundations/RelationalStructure.agda:116]: Use \exists

File: Cubical/Structures/Ring.agda [hunk 2]
```diff
@@ -151,10 +152,12 @@ record RingEquiv (R S : Ring {ℓ}) : Type ℓ where
 
 module RingΣTheory {ℓ} where
 
-  open Macro ℓ (recvar (recvar var) , var , recvar (recvar var)) public renaming
-    ( structure to RawRingStructure
-    ; equiv     to RawRingEquivStr
-    ; univalent to rawRingUnivalentStr )
+  RawRingStructure = λ (X : Type ℓ) → (X → X → X) × X × (X → X → X)
```
Comment (mortberg) [inline Cubical/Structures/Ring.agda:155]: Is this kind of change already done for other structures?

Comment (ecavallo) [reply Cubical/Structures/Ring.agda:155]: Yeah, I just missed this one.

**Pull request merged**